### ISSUE TITLE
401 Missing or invalid authentication.

### ISSUE THREAD

**Issue by fdebrus** [fdebrus/hayward-ha#3]
After the token expiration time, token is not getting refreshed leading to 401 error.
You can only restart HA to resolve.

regular way of refreshing token will not work here. open for idea and suggestion.

**Comment by fdebrus** [comment]
some progress on https://github.com/fdebrus/hayward-ha/tree/0.0.5.4
enable debuglog and monitor for token check and token refresh events

**Comment by Hans1205** [comment]
Which version is without logging /debugging? After using 0.0.5.3 microSD Card on RASPI was 100% full in hours ...

**Comment by fdebrus** [comment]
if "enable debug logging" is not activated at integration level, you should not get a lot of message in the log. if you could report to me what was in the log ? so I can try to find the source.

![image](https://github.com/fdebrus/hayward-ha/assets/33791533/2f0a7c24-34d6-41aa-b859-2d506a74896a)

**Comment by Hans1205** [comment]
No, unfortunately I can't. I have already deleted the log and commented out everything that is related to logging in aquarite.py.

However, the disk space is still much fuller than yesterday. Before setting up Aquarite I had 59% occupied. This morning at 9 o'clock 100% and after deleting the homeassistant.log still 90%

Von meinem iPhone gesendet

Hans Moggert

Am 22.04.2024 um 16:09 schrieb fdebrus ***@***.***>:

﻿

Could you tell me what message was filling the log ? without debugging enabled, it should not happen

**Comment by fdebrus** [comment]
ok, I've opened a new issue to track that disk getting full https://github.com/fdebrus/hayward-ha/issues/4

**Comment by fdebrus** [comment]
Refresh of the token is working now, but I still need to work on the firestore document handling after refreshing the token

aka, if token expires it get refreshed and you can keep sending command to pool, but pool updates are not reflected in HA; WIP

**Comment by fdebrus** [comment]
I've changed the subscription to firestore to re-auth in case a new token is created. code below

But I will change this code moving away from realtime updates towards scheduled refresh in the next release. This generates too many datapoints

```    
    async def subscribe(self, pool_id, handler) -> None:
        await self.ensure_active_token()
        doc_ref = self.client.collection("pools").document(pool_id)
        def on_snapshot(doc_snapshot, changes, read_time):
            """Handles document snapshots."""
            for doc in doc_snapshot:
                handler(doc)
        doc_ref.on_snapshot(on_snapshot)
        self.handlers.append(handler)

    async def auto_resubscribe(self, pool_id, handler):
        try:
            await self.subscribe(pool_id, handler)
        except Exception as e:
            _LOGGER.error("Subscription failed, attempting to resubscribe...")
            await asyncio.sleep(10)
            await self.auto_resubscribe(pool_id, handler) 
```

**Comment by fdebrus** [comment]
solved in 0.0.6

**Comment by fdebrus** [comment]
While the command are sent to the pool after token refresh, the pool updates are no longer received by the integration following a token refresh. I suspect the coordinator not to receive the updated client following token refresh. to be investigated...

**Comment by fdebrus** [comment]
closed in 0.0.7
